plugins/last-fm.py

```diff
@@ -0,0 +1,237 @@
+    if response.status_code == 200:
+        try:
+            track = response.json()['recenttracks']['track'][0]
+            return track
+        except IndexError:
+            pass
+
+
+def _pull_track_tags(api_key, track_title, track_artist):
+    url = api_url+f'&method=track.gettoptags&track={track_title}'
+    url += f'&artist={track_artist}&api_key={api_key}'
+    response = requests.get(url)
+    if response.status_code == 200:
+        try:
+            tags_json = response.json()['toptags']['tag'][0:5]
+        except IndexError:
+            tags_json = response.json()['toptags']['tag']
+
+        if len(tags_json) < 1:
+            return 'none'
+
+        tags = []
+        for tag in tags_json:
+            tags.append(tag['name'])
+        tags = ', '.join(tags)
+        return tags
+
+
+def _get_when_played(track):
+    """returns the time difference between right now and the time the
+    track was played in text format"""
+
+    """i decided to not include this in the output since the API
+     is either buggy, or is reacting too slowly to changes. i.e a track
+     i had just scrobbled shows up as 'played one hour ago' """
+    try:
+        listened = datetime.strptime(track['date']['#text'], '%d %b %Y, %H:%M')
+        curtime = datetime.fromtimestamp(time.time())
+        curtime = datetime.strptime(str(curtime), '%Y-%m-%d %H:%M:%S.%f')
+        diff = str((curtime - listened))
+        listened = diff.split('.')[0]
+        if listened[0:2] == '-1':
+            listened = listened[8:-6] + ' seconds'
+        elif listened[0] == '0':
+            if listened[2:4] == '01':
+                listened = listened[3:-3] + ' minute'
+            elif listened[2:3] == '0':
+                if listened[2:4] == '00':
+                    listened = '1 minute'
+                else:
+                    listened = listened[3:-3] + ' minutes'
+            else:
+                listened = listened[2:-3] + ' minutes'
+        else:
+            if listened[-8:-6] == ' 1' or listened[0:2] == '1:':
+                listened = listened[:-6] + ' hour'
+            else:
+                listened = listened[:-6] + ' hours'
+        return listened
+    except KeyError:
+        pass
+
+
+def _pull_track_plays(api_key, track_title, track_artist, lastfm_nick):
+    url = api_url+f'&method=track.getinfo&username={lastfm_nick}'
+    url += f'&track={track_title}&artist={track_artist}'
+    url += f'&api_key={api_key}'
+    response = requests.get(url)
+    if response.status_code == 200:
+        try:
+            playcount_json = response.json()['track']['userplaycount']
+            return str(playcount_json)
+        except:
+            return 'unknown'
+
+
+def _output_nowplaying(client, data, lastfm_nick,
+                       title, artist,album, plays, tags):
+    out = f'{lastfm_nick} is listening to "{title}" by {artist} from '
+    if album == 'an unknown album':
+        out += 'an unknown album'
```

Comment:
Could use `out = f'{out} an unknown album'`

Reply:
fstrings are actually a *tiny* bit slower than just doing out += var, so i see no point in using them unless i need to insert a variable. I know that the difference is minute, but here is some proof:

appending via +=:
```
#test.py
out = ''

for i in range(0, 100000):
    out += 'a'
```
```
$time python test.py
real	0m0.041s
user	0m0.037s
sys	0m0.004s
```
fstrings:
```
#test.py
out = ''

for i in range(0, 100000):
    out = f'{out}a'
```
```
$time python test.py
real	0m0.247s
user	0m0.213s
sys	0m0.034s
```